File: 4kyu/human_readable_duration_format.cpp
#include <string>

std::string format_duration(int n) {
    if (n == 0)
        return "now";
    int y = n / 31536000, yr = n % 31536000;
    int d = yr / 86400, dr = yr % 86400;
    int h = dr / 3600, hr = dr % 3600;
    int m = hr / 60, s = hr % 60;
    int time[5] = {y, d, h, m, s};
    std::string duration;
    std::string fmt[5] = {"year", "day", "hour", "minute", "second"};
    int count = -1;
    for (int i = 0; i < 5; i++)
        if (time[i])
            count++;
    bool first = true;
    for (int i = 0; i < 5; i++) {
        if (time[i]) {
            duration += std::to_string(time[i]) + " " + fmt[i];
            if (time[i] > 1)
                duration += 's';
            count--;
            if (count > 0)
                duration += ", ";
            if (!count)
                duration += " and ";
        }
    }
    return duration;
}

File: 4kyu/human_readable_duration_format.py
def format_duration(n: int) -> str:
    count = {k: 0 for k in 'year day hour minute second'.split()}
    count['year'], n = divmod(n, 31536000)
    count['day'], n = divmod(n, 86400)
    count['hour'], n = divmod(n, 3600)
    count['minute'], count['second'] = divmod(n, 60)
    ans = ''
    c = {k + 's' if v > 1 else k: v for k, v in count.items() if v}
    nz = len(c)
    for k, v in c.items():
        nz -= 1
        separator = ', ' if nz > 1 else ' and ' if nz else ''
        ans += f'{v} {k}' + separator
    return ans if ans else 'now'
